--- baresip-4.5.0/include/baresip.h	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/include/baresip-new.h	2026-03-07 07:24:23.141954857 +0400
@@ -291,6 +291,9 @@
 enum sip_transp call_transp(const struct call *call);
 enum sdp_neg_state call_sdp_neg_state(const struct call *call);
 bool call_sdp_change_allowed(const struct call *call);
+void call_xuser_data_set(struct call *call, void *data, bool destroy);
+void *call_xuser_data_get(struct call *call);
+
 
 /**
  * Command parameter parse helpers
@@ -1016,7 +1019,8 @@
 void ua_remove_extension(struct ua *ua, const char *extension);
 bool ua_req_allowed(const struct ua *ua, const struct sip_msg *msg);
 bool ua_req_check_origin(const struct ua *ua, const struct sip_msg *msg);
-
+void ua_xuser_data_set(struct ua *ua, void *data, bool destroy);
+void *ua_xuser_data_get(struct ua *ua);
 
 /* One instance */
 int  ua_init(const char *software, bool udp, bool tcp, bool tls);
@@ -1030,6 +1034,7 @@
 void uag_enable_sip_trace(bool enable);
 int  uag_reset_transp(bool reg, bool reinvite);
 void uag_set_sub_handler(sip_msg_h *subh);
+void uag_set_custom_sip_handler(sip_msg_h *handler);
 int  uag_set_extra_params(const char *eprm);
 int  uag_enable_transport(enum sip_transp tp, bool en);
 struct ua   *uag_find(const struct pl *cuser);
@@ -1474,7 +1479,8 @@
 struct config_audio *audio_config(struct audio *au);
 bool audio_txtelev_empty(const struct audio *au);
 void audio_call_telev_handler(const struct audio *au, int key, bool end);
-
+void audio_xuser_data_set(struct audio *au, void *data, bool destroy);
+void *audio_xuser_data_get(struct audio *au);
 
 /*
  * Video stream
@@ -1512,6 +1518,8 @@
 const struct vidcodec *video_codec(const struct video *vid, bool tx);
 void video_sdp_attr_decode(struct video *v);
 void video_req_keyframe(struct video *vid);
+void video_xuser_data_set(struct video *vid, void *data, bool destroy);
+void *video_xuser_data_get(struct video *vid);
 
 double video_calc_seconds(uint64_t rtp_ts);
 double video_timestamp_to_seconds(uint64_t timestamp);
--- baresip-4.5.0/src/core.h	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/core-new.h	2026-02-20 13:53:45.496352551 +0400
@@ -391,6 +391,7 @@
 	uint32_t transports;           /**< Supported transports mask       */
 	bool delayed_close;            /**< Module will close SIP stack     */
 	sip_msg_h *subh;               /**< Subscribe handler               */
+	sip_msg_h *xsiph;              /**< akstel sip extra handler        */
 	ua_exit_h *exith;              /**< UA Exit handler                 */
 	bool nodial;                   /**< Prevent outgoing calls          */
 	void *arg;                     /**< UA Exit handler argument        */
--- baresip-4.5.0/src/uag.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/uag-new.c	2026-03-05 03:38:20.932187439 +0400
@@ -204,15 +204,21 @@
 
 	(void)arg;
 
+    if (uag.xsiph) {
+         if(uag.xsiph(msg, NULL)) {
+            return true;
+        }
+    }
+
 	if (pl_strcmp(&msg->met, "OPTIONS") &&
 	    pl_strcmp(&msg->met, "REFER"))
 		return false;
 
-	ua = uag_find_msg(msg);
-	if (!ua) {
-		(void)sip_treply(NULL, uag_sip(), msg, 404, "Not Found");
-		return true;
-	}
+    ua = uag_find_msg(msg);
+    if (!ua) {
+        (void)sip_treply(NULL, uag_sip(), msg, 404, "Not Found");
+        return true;
+    }
 
 	if (!ua_req_check_origin(ua, msg)) {
 		(void)sip_treply(NULL, uag_sip(), msg, 403, "Forbidden");
@@ -1307,3 +1313,12 @@
 {
 	return uag.subh;
 }
+
+/**
+ * akstel.org
+ * this handler is procesing in request_handler()
+ */
+void uag_set_custom_sip_handler(sip_msg_h *handler)
+{
+	uag.xsiph = handler;
+}
--- baresip-4.5.0/src/ua.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/ua-new.c	2026-03-04 23:40:53.351792307 +0400
@@ -30,6 +30,8 @@
 	struct list custom_hdrs;     /**< List of outgoing headers           */
 	char *ansval;                /**< SIP auto answer value              */
 	struct sa dst;               /**< Current destination address        */
+	void    *xuser_data;         /** akstel.org */
+	bool    xud_destroy;         /** akstel.org */
 };
 
 struct ua_xhdr_filter {
@@ -61,6 +63,10 @@
 	mem_deref(ua->ansval);
 	mem_deref(ua->acc);
 
+	if(ua->xud_destroy) {
+        mem_deref(ua->xuser_data);
+	}
+
 	if (uag_delayed_close() && list_isempty(uag_list())) {
 		sip_close(uag_sip(), false);
 	}
@@ -2355,3 +2361,16 @@
 
 	return false;
 }
+
+/* akstel.org */
+void ua_xuser_data_set(struct ua *ua, void *data, bool destroy) {
+	if (!ua) return;
+    ua->xuser_data = data;
+    ua->xud_destroy = destroy;
+}
+
+/* akstel.org */
+void *ua_xuser_data_get(struct ua *ua) {
+    return ua ? ua->xuser_data : NULL;
+}
+
--- baresip-4.5.0/src/call.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/call-new.c	2026-03-04 23:26:55.689887299 +0400
@@ -84,6 +84,8 @@
 	bool use_video;
 	bool use_rtp;
 	struct pl *user_data;      /**< User data related to the call       */
+	void    *xuser_data;       /** akstel.org */
+	bool    xud_destroy;       /** if true will be destroyed */
 };
 
 
@@ -363,6 +365,10 @@
 	mem_deref(call->acc);
 	mem_deref(call->user_data);
 
+	if(call->xud_destroy) {
+        mem_deref(call->xuser_data);
+	}
+
 	list_flush(&call->custom_hdrs);
 }
 
@@ -3454,3 +3460,16 @@
 		|| (sdp_state == SDP_NEG_NONE
 		|| sdp_state == SDP_NEG_REMOTE_OFFER);
 }
+
+/* akstel.org */
+void call_xuser_data_set(struct call *call, void *data, bool destroy) {
+	if (!call) return;
+
+    call->xuser_data = data;
+    call->xud_destroy = destroy;
+}
+
+/* akstel.org */
+void *call_xuser_data_get(struct call *call) {
+    return call ? call->xuser_data : NULL;
+}
--- baresip-4.5.0/src/audio.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/audio-new.c	2026-03-07 07:20:55.294457172 +0400
@@ -133,6 +133,8 @@
 	audio_level_h *levelh;        /**< Audio level handler             */
 	audio_err_h *errh;            /**< Audio error handler             */
 	void *arg;                    /**< Handler argument                */
+	void *xuser_data;             /** akstel.org */
+	bool xud_destroy;             /** if true will be destroyed */
 };
 
 
@@ -214,6 +216,10 @@
 	mem_deref(a->telev);
 	mem_deref(a->aur);
 
+	if(a->xud_destroy) {
+        mem_deref(a->xuser_data);
+	}
+
 	mem_deref(a->tx.mtx);
 }
 
@@ -1932,3 +1938,16 @@
 {
 	return au ? &au->cfg : NULL;
 }
+
+/* akstel.org */
+void audio_xuser_data_set(struct audio *au, void *data, bool destroy) {
+	if (!au) return;
+
+    au->xuser_data = data;
+    au->xud_destroy = destroy;
+}
+
+/* akstel.org */
+void *audio_xuser_data_get(struct audio *au) {
+    return au ? au->xuser_data : NULL;
+}
--- baresip-4.5.0/src/video.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/video-new.c	2026-03-07 07:26:43.780204441 +0400
@@ -169,6 +169,8 @@
 	bool nack_pli;          /**< Send NACK/PLI to peer                */
 	video_err_h *errh;      /**< Error handler                        */
 	void *arg;              /**< Error handler argument               */
+	void *xuser_data;       /** akstel.org */
+	bool xud_destroy;       /** if true will be destroyed */
 };
 
 
@@ -1965,3 +1967,16 @@
 	vid->vtx.picup = true;
 	mtx_unlock(vid->vtx.lock_enc);
 }
+
+/* akstel.org */
+void video_xuser_data_set(struct video *vid, void *data, bool destroy) {
+	if (!vid) return;
+
+    vid->xuser_data = data;
+    vid->xud_destroy = destroy;
+}
+
+/* akstel.org */
+void *video_xuser_data_get(struct video *vid) {
+    return vid ? vid->xuser_data : NULL;
+}
